updated to function component

keisyrzk opened the pull request:
### Details
Updated the component so it is now built with a `function`.

### Fixed Issues
$ https://github.com/Expensify/App/issues/16119


### Tests
1. Log in to the app
2. click avatar 
3. click avatar placed over user's email
4a. click every row in the list and go back
4b. click on the avatar so the bottom sheet appears
5. click "upload photo"
6a. click on the screen outside the sheet - check if closes without issues
6b. click on the "gallery" option
7. Gallery - pick a photo
8. Edit photo and "save"
9. When taken back to profile view click again on the avatar (already set) so the bottom sheet appears with new set of options
10a. click on the screen outside the sheet - check if closes without issues
10b. click "view photo" and then go back
10c. click "upload photo" and repeat the proccess of picking a photo
10d. click "remove photo" - see if avatar changed to default one
11. exit the view to the app's main view and return back here to verify if the default avatar persists on every screen
12. do the same with the avatar set with some photo


- :white_check_mark: Verify that no errors appear in the JS console

### Offline tests
NA

### QA Steps
The same as described in "Tests" section.

- [ x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Web</summary>
https://github.com/Expensify/App/assets/18525360/63ad1dc3-6658-4ea3-b1ce-f1d9dc2d89ee
https://github.com/Expensify/App/assets/18525360/3c79afc2-beff-4b3c-9211-c5e03389727d
</details>

<details>
<summary>Mobile Web - Chrome</summary>
https://github.com/Expensify/App/assets/18525360/191935fc-244a-4786-a5f5-60a8f83002bc
</details>

<details>
<summary>Mobile Web - Safari</summary>
https://github.com/Expensify/App/assets/18525360/fbeadfcb-c3b7-4c7a-b7bf-aeea8334951d
</details>

<details>
<summary>Desktop</summary>
https://github.com/Expensify/App/assets/18525360/e313dbec-f336-4cf0-b790-2a3b19d491d5
</details>

<details>
<summary>iOS</summary>
https://github.com/Expensify/App/assets/18525360/b4b57df2-48dd-4b84-b97b-0269b32fcfa9
</details>

<details>
<summary>Android</summary>
https://github.com/Expensify/App/assets/18525360/adfc88ae-db17-4188-8275-33e8d5e73e16
</details>

<details>
<summary>Extra - workspace avatar</summary>
Uploading workspaceAvatar.mov…
</details>

keisyrzk committed:
updated to function componenet

gedu commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -1,6 +1,6 @@
 import _ from 'underscore';
-import React from 'react';
-import {View} from 'react-native';
+import React, { useState, useRef, useEffect } from 'react';
+import { View } from 'react-native';
```
It shouldn't be a space between `{ }`

gedu commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
-            imageName: '',
-            imageUri: '',
-            imageType: '',
+function AvatarWithImagePicker(props) {
```
You can use [prop destructuring](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#destructuring)

gedu commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
-            imageUri: '',
-            imageType: '',
+function AvatarWithImagePicker(props) {
+    const animation = new SpinningIndicatorAnimation();
```
this will be recreated on every new frame, I think is better to use `useRef`
```
    const animation = useRef(new SpinningIndicatorAnimation()).current;
```

gedu commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
+    const anchorRef = useRef();
+
+    useEffect(() => {
+        if (props.isUploading) {
```
I'm seeing the start, but not seeing the `animation.stop`, the stop from the return only gets executed when the Component is unmounted, but seems that we should stop the animation when the `isUploading` change

keisyrzk replied:
it is just beneath the line you commented

`    useEffect(() => {
        if (isUploading) {
            animation.start();
        }
        
        return () => {
            animation.stop();
        };
    }, [isUploading]);`

gedu commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
+        if (!props.isFocused) {
+            setError(null, {});
         }
+    }, [props.isFocused]);
```
if you don't add `setError` as dependency, please add a comment explaining why

rezkiy37 replied:
Either you can use `useCallback` for `setError` to memorise it and safely add to the deps. arrays.

fabioh8010 commented on ios/Podfile.lock:
Please revert the changes on this file, they shouldn't have changed in this PR.

keisyrzk replied:
done

fabioh8010 commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
-     */
-    isValidExtension(image) {
-        const {fileExtension} = FileUtils.splitExtensionFromFileName(lodashGet(image, 'name', ''));
+    const isValidExtension = (image) => {
```
```suggestion
    /**
     * Check if the attachment extension is allowed.
     *
     * @param {Object} image
     * @returns {Boolean}
     */
    const isValidExtension = (image) => {
```

let's keep the comments

fabioh8010 commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
 
-        this.animation.start();
-    }
+    const setError = (error, phraseParam) => {
```
```suggestion
    /**
     * @param {String} error
     * @param {Object} phraseParam
     */
    const setError = (error, phraseParam) => {
```

let's keep the comments

fabioh8010 commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
-    isValidSize(image) {
+    };
+
+    const isValidSize = (image) => {
```
```suggestion
    /**
     * Check if the attachment size is less than allowed size.
     *
     * @param {Object} image
     * @returns {Boolean}
     */
    const isValidSize = (image) => {
```

let's keep the comments

fabioh8010 commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
-            this.setError('avatarWithImagePicker.notAllowedExtension', {allowedExtensions: CONST.AVATAR_ALLOWED_EXTENSIONS});
+    };
+
+    const isValidResolution = async (image) => {
```
```suggestion
    /**
     * Check if the attachment resolution matches constraints.
     *
     * @param {Object} image
     * @returns {Promise}
     */
    const isValidResolution = async (image) => {
```

fabioh8010 commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
+            resolution.width <= CONST.AVATAR_MAX_WIDTH_PX;
+    };
+
+    const showAvatarCropModal = async (image) => {
```
```suggestion
    /**
     * Validates if an image has a valid resolution and opens an avatar crop modal
     *
     * @param {Object} image
     */
    const showAvatarCropModal = async (image) => {
```

fabioh8010 commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
-                imageUri: image.uri,
-                imageName: image.name,
-                imageType: image.type,
+        const isValidRes = await isValidResolution(image);
```
async/await is forbidden -> https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#esnext-are-we-allowed-to-use-insert-new-language-feature-why-or-why-not

gedu commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
+        setIsAvatarCropModalOpen(true);
+        setValidationError(null);
+        setPhraseParam({});
+        setIsMenuVisible(false);
+        setImageUri(image.uri);
+        setImageName(image.name);
+        setImageType(image.type);
```
I'm seeing all of this are being used all together, I would create a single `useState` which uses an object, because those `setXXX` being inside of an async function React-Native can't batch the state updates, so right now will be 7 individual updates.

keisyrzk replied:
U R right, except the `setIsMenuVisible` - updated

gedu commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
-            <View style={[styles.alignItemsCenter, ...additionalStyles]}>
-                <View style={[styles.pRelative, styles.avatarLarge]}>
+        setIsAvatarCropModalOpen(true);
+        setValidationError(null);
+        setPhraseParam({});
```
As I wrote below, I would merge these 2 states into one, because they are being used together. doing that we can reduce at least 3 states updates

gedu commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
+    };
+
+    const DefaultAvatar = props.DefaultAvatar;
+    const additionalStyles = _.isArray(props.style) ? props.style : [props.style];
```
you can use `StyleSheet.flatten`
```
const additionalStyles = StyleSheet.flatten(style);
```

keisyrzk replied:
actually the change you suggested causes an error "not iterable" so I keep my version

gedu commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,166 +116,118 @@ const defaultProps = {
+    const additionalStyles = _.isArray(props.style) ? props.style : [props.style];
+
+    return (
+        <View style={[styles.alignItemsCenter, ...additionalStyles]}>
```
to avoid the spread operator I would add all into the `flatten`
```
const additionalStyles = StyleSheet.flatten([styles.alignItemsCenter, style]);
```

keisyrzk replied:
ok, but would like to hear why you believe it is better?

eVoloshchak replied:
+1 for using `Stylesheet.flatten`, that would mitigate the need for `additionalStyles` const

gedu commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -293,92 +245,91 @@ class AvatarWithImagePicker extends React.Component {
-        );
-    }
+        </View>
+    );
```
replace the `withLocalize` to `useLocalize` hook
add the `displayName` [prop](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#function-component-style).

rezkiy37 reviewed (commented):
LGTM, after resolving comments.

keisyrzk committed:
updates

keisyrzk committed:
comments

keisyrzk commented:
I have read the CLA Document and I hereby sign the CLA

keisyrzk committed:
pr updates

blazejkustra commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -293,95 +317,94 @@ class AvatarWithImagePicker extends React.Component {
 AvatarWithImagePicker.defaultProps = defaultProps;
 
-export default compose(withLocalize, withNavigationFocus)(AvatarWithImagePicker);
+export default withNavigationFocus(AvatarWithImagePicker);
```
Use `const isFocused = useIsFocused();` hook instead of `withNavigationFocus` HOC.

Most likely it will cause app to crash due to this [bug](https://github.com/Expensify/App/pull/23160#issuecomment-1673302723).

keisyrzk replied:
yes, will keep it

blazejkustra reviewed (changes requested):
Vit asked me to review once PR is up, I left some notes from the previous PR that I worked on.

fabioh8010 commented on ios/Podfile.lock:
```diff
@@ -783,7 +783,7 @@ PODS:
   - RNReactNativeHapticFeedback (1.14.0):
     - React-Core
-  - RNReanimated (3.5.4):
+  - RNReanimated (3.4.0):
```
Let's remove this file @keisyrzk

keisyrzk committed:
remove podfile.lock

keisyrzk commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -12,7 +12,7 @@ import themeColors from '../styles/themes/default';
+import useLocalize from '../hooks/useLocalize';
 import variables from '../styles/variables';
 import CONST from '../CONST';
 import SpinningIndicatorAnimation from '../styles/animation/SpinningIndicatorAnimation';
```
i leave it as it is as your comment is unclear and I was already asked to do as it is now

blazejkustra replied:
@keisyrzk You shouldn't resolve comments because something is 'unclear'. Instead please ask a question or just ask to make the comment more clear.

The thing is that `SpinningIndicatorAnimation` in this component is not really used (it's used only in lifecycle methods):
```ts
 const animation = useRef(new SpinningIndicatorAnimation()).current;

    useEffect(() => {
        if (isUploading) {
            animation.start();
        }

        return () => {
            animation.stop();
        };
    }, [isUploading]);
```

> I was already asked to do as it is now

What do you mean? Who did it and what's the reasoning behind it?

keisyrzk replied:
ok i misunderstood you

blazejkustra replied:
No problem 😄

keisyrzk committed:
updates pr

keisyrzk committed:
restore podfile.lock

keisyrzk committed:
podfile.lock update content

keisyrzk committed:
podfile.lock empty space missing

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -114,168 +113,229 @@ const defaultProps = {
+    anchorPosition, 
+    anchorAlignment, 
+    onImageSelected, 
+    editorMaskImage  }) {
```
```suggestion
    editorMaskImage,
}) {
```

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -114,168 +113,229 @@ const defaultProps = {
-        return (
-            <View style={[styles.alignItemsCenter, ...additionalStyles]}>
-                <View style={[styles.pRelative, styles.avatarLarge]}>
+        /**
```
A small misalignment issue

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -114,168 +113,229 @@ const defaultProps = {
+     * @param {Function} openPicker
+     * @returns {Array}
+     */
+        const createMenuItems = (openPicker) => {
```
Also should be one `Tab` to the left

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -114,168 +113,229 @@ const defaultProps = {
+        
+    return (
+        <View style={[styles.alignItemsCenter, ...additionalStyles]}>
+                            <View style={[styles.pRelative, styles.avatarLarge]}>
```
This `View` isn't aligned properly

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -293,95 +353,73 @@ class AvatarWithImagePicker extends React.Component {
-                                                    return;
-                                                }
+                            <AttachmentPicker>
+                            {({openPicker}) => {
```
This should be one `Tab` to the right

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -293,95 +353,73 @@ class AvatarWithImagePicker extends React.Component {
 }
 
 AvatarWithImagePicker.propTypes = propTypes;
 AvatarWithImagePicker.defaultProps = defaultProps;
```
> When writing a function component you must ALWAYS add a displayName property and give it the same value as the name of the component (this is so it appears properly in the React dev tools)

https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#function-component-style

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -293,95 +353,73 @@ class AvatarWithImagePicker extends React.Component {
 AvatarWithImagePicker.defaultProps = defaultProps;
 
-export default compose(withLocalize, withNavigationFocus)(AvatarWithImagePicker);
+export default withNavigationFocus(AvatarWithImagePicker);
```
```suggestion
export default withNavigationFocus(AvatarWithImagePicker);

```

eVoloshchak reviewed (changes requested):
Apologies, I meant to 'Request Changes' instead of approving
@deetergp, ignore this please

keisyrzk committed:
Update src/components/AvatarWithImagePicker.js

Co-authored-by: Eugene Voloshchak <copyreading@gmail.com>

keisyrzk committed:
pr updates

blazejkustra commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -114,168 +113,230 @@ const defaultProps = {
-        this.isValidResolution(image).then((isValidResolution) => {
-            if (!isValidResolution) {
-                this.setError('avatarWithImagePicker.resolutionConstraints', {
+    function showAvatarCropModal(image) {
```
For consistency:
```suggestion
    const showAvatarCropModal = (image) => {
```

blazejkustra commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -114,168 +113,230 @@ const defaultProps = {
-            if (!isValidResolution) {
-                this.setError('avatarWithImagePicker.resolutionConstraints', {
+    function showAvatarCropModal(image) {
+        const handleResolutionValidation = (isValidRes) => {
```
I would keep it simple:
```ts
   const showAvatarCropModal = (image) => {
        if (!isValidExtension(image)) {
            setError('avatarWithImagePicker.notAllowedExtension', {allowedExtensions: CONST.AVATAR_ALLOWED_EXTENSIONS});
            return;
        }
        if (!isValidSize(image)) {
            setError('avatarWithImagePicker.sizeExceeded', {maxUploadSizeInMB: CONST.AVATAR_MAX_ATTACHMENT_SIZE / (1024 * 1024)});
            return;
        }

        isValidResolution(image).then((isValid) => {
            if (!isValid) {
                setError('avatarWithImagePicker.resolutionConstraints', {
                    minHeightInPx: CONST.AVATAR_MIN_HEIGHT_PX,
                    minWidthInPx: CONST.AVATAR_MIN_WIDTH_PX,
                    maxHeightInPx: CONST.AVATAR_MAX_HEIGHT_PX,
                    maxWidthInPx: CONST.AVATAR_MAX_WIDTH_PX,
                });
                return;
            }

            setIsAvatarCropModalOpen(true);
            setError(null, {});
            setIsMenuVisible(false);
            setImageData({
                uri: image.uri,
                name: image.name,
                type: image.type
            });
        });
    };
```

blazejkustra commented:
Also, I added comment to a resolved discussion [here](https://github.com/Expensify/App/pull/28625#discussion_r1348412945)

keisyrzk committed:
updates

eVoloshchak commented on ios/Podfile.lock:
```diff
@@ -1311,4 +1311,4 @@ SPEC CHECKSUMS:
 PODFILE CHECKSUM: 2daf34c870819a933f3fefe426801d54b2ff2a14
 
-COCOAPODS: 1.12.1
+COCOAPODS: 1.12.1
```
These changes are unrelated to this PR

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,112 +114,115 @@ const defaultProps = {
+    onImageSelected, 
+    editorMaskImage,
+}) {
+
```
Empty line isn't needed here

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,112 +114,115 @@ const defaultProps = {
+    };
 
-    /**
+     /**
```
This is misaligned

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -230,52 +231,85 @@ class AvatarWithImagePicker extends React.Component {
+    return (
+        <View style={[styles.alignItemsCenter, ...additionalStyles]}>
+            <View style={[styles.pRelative, styles.avatarLarge]}>
                     <OfflineWithFeedback
```
This is misaligned

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -293,95 +327,74 @@ class AvatarWithImagePicker extends React.Component {
-                                                if (Browser.isSafari()) {
-                                                    return;
-                                                }
+                    {({show}) => (
```
This is misaligned

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,112 +114,115 @@ const defaultProps = {
+    const setError = (error, phraseParam) => {
+        setErrorData({
+            validationError: error,
+            phraseParam: phraseParam,
```
```suggestion
            phraseParam,
```

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,112 +114,115 @@ const defaultProps = {
      * @returns {Boolean}
      */
-    isValidSize(image) {
+    const isValidSize = (image) => {
```
```suggestion
    const isValidSize = (image) => image && lodashGet(image, 'size', 0) < CONST.AVATAR_MAX_ATTACHMENT_SIZE;
```

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,112 +114,115 @@ const defaultProps = {
-        );
-    }
+    const isValidResolution = async (image) => {
+        const resolution = await getImageResolution(image);
```
https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#esnext-are-we-allowed-to-use-insert-new-language-feature-why-or-why-not

> Async/Await - Use the native Promise instead

<img width="723" alt="image" src="https://github.com/Expensify/App/assets/9059945/a5f2ecda-bebc-4acd-9906-0491e744d470">

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,112 +114,115 @@ const defaultProps = {
-        if (!this.props.isUploading) {
-            return;
+function AvatarWithImagePicker({ 
+    isUploading, 
```
<img width="550" alt="image" src="https://github.com/Expensify/App/assets/9059945/99b3b57c-52b5-41e8-aa3c-4b699b3fbeb9">

eVoloshchak replied:
We're not using `compose` anymore, so the same linter warning

<img width="732" alt="image" src="https://github.com/Expensify/App/assets/9059945/6cb6aef3-5c8a-411a-9909-26e39c6e32d6">

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,112 +114,115 @@ const defaultProps = {
+
+    useEffect(() => {
+        // check if the component is no longer focused, then reset the error.
+        if (!isFocused) {
```
Linter error

<img width="828" alt="image" src="https://github.com/Expensify/App/assets/9059945/a927dda7-1baa-422c-8636-73d5448b3823">

keisyrzk committed:
lint updates

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -91,137 +89,141 @@ const propTypes = {
+        type: ''
+    });
+    const anchorRef = useRef();
+    const { translate } = useLocalize();
```
```suggestion
    const {translate} = useLocalize();
```

And same with all other occurrences of space before and after brackets

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -91,137 +89,141 @@ const propTypes = {
+    const { translate } = useLocalize();
+    
+    useEffect(() => {
+        // If the component is still focused, don't proceed further.
```
```suggestion
```

This is redundant since there's a comment below

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -91,137 +89,141 @@ const propTypes = {
-    isValidSize(image) {
-        return image && lodashGet(image, 'size', 0) < CONST.AVATAR_MAX_ATTACHMENT_SIZE;
-    }
+    * Check if the attachment size is less than allowed size.
```
This comment is misaligned

eVoloshchak commented on ios/Podfile.lock:
This isn't resolved, these changes are unrelated to the PR

keisyrzk committed:
pr empty sapces fix

eVoloshchak commented on ios/Podfile.lock:
Do not mark this conversation as `Resolved`, these changes shouldn't be here, they're unrelated to the PR

eVoloshchak replied:
I can still see changes applied to 2 files instead of 1, one of them is Podfile.lock (there is a new line added at the end of file)
<img width="187" alt="image" src="https://github.com/Expensify/App/assets/9059945/19cfb349-afe2-471e-ac9c-6e42aa75481c">

eVoloshchak commented:
There are a couple of linter errors still

<img width="658" alt="image" src="https://github.com/Expensify/App/assets/9059945/f45ded27-8d97-4c09-a5d8-fff9a75c7d55">

They're should be highlighted in IDE, but you can also run linter from the command line. To do that

1. Add a new line containing `**/vendor/*` [eslintignore](https://github.com/Expensify/App/blob/main/.eslintignore)
2. Run `npm run lint` from command line

keisyrzk committed:
lint fix

blazejkustra commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -12,10 +12,9 @@ import themeColors from '../styles/themes/default';
+import useLocalize from '../hooks/useLocalize';
 import variables from '../styles/variables';
 import CONST from '../CONST';
-import SpinningIndicatorAnimation from '../styles/animation/SpinningIndicatorAnimation';
```
@keisyrzk Remove `SpinningIndicatorAnimation` file, it isn't used in the codebase.

keisyrzk replied:
ok done

keisyrzk committed:
remove SpinningIndicatorAnimation file

eVoloshchak reviewed (commented):
Looks good except for the unrelated changes in `Podfile.lock`

keisyrzk commented:
> Looks good except for the unrelated changes in `Podfile.lock`

I have not changed the podfile.lock, I kept the original one but had to push the changes as accidentally pushed podfile.lock earlier.

eVoloshchak commented:
@keisyrzk

- just undo the changes to the `podfile.lock` (i.e. add the new line at the end of the file)
- pull the latest main

keisyrzk commented:
> @keisyrzk
> 
> * just undo the changes to the `podfile.lock` (i.e. add the new line at the end of the file)
> * pull the latest main

The file has been already reverted to the original.

keisyrzk committed:
merge latest main

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,112 +109,114 @@ const defaultProps = {
-    }
-
-    /**
+function AvatarWithImagePicker({
```
We don't need `defaultProps` anymore, let's just set default values here as per https://expensify.slack.com/archives/C01GTK53T8Q/p1688785407639549

> destructuring props in function declarations is now documented as a best practice, and should be used instead of default props going forward. More info available at https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#destructuring

eVoloshchak replied:
Please take a look at this comment

keisyrzk replied:
@eVoloshchak we need to keep the `defaultProps` as in some cases like this one it is not possible to remove the `defaultProps` because of the linter errors. I have consulted this and been asked to keep the `defaultProps` as they should be kept in JS files for know.

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,112 +109,114 @@ const defaultProps = {
-    isValidExtension(image) {
-        const {fileExtension} = FileUtils.splitExtensionFromFileName(lodashGet(image, 'name', ''));
+    const isValidExtension = (image) => {
+        const { fileExtension } = FileUtils.splitExtensionFromFileName(lodashGet(image, 'name', ''));
```
```suggestion
        const {fileExtension} = FileUtils.splitExtensionFromFileName(lodashGet(image, 'name', ''));
```

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -116,112 +109,114 @@ const defaultProps = {
-                resolution.width <= CONST.AVATAR_MAX_WIDTH_PX,
+    const isValidResolution = (image) => 
+    getImageResolution(image)
+        .then(({ height, width }) => 
```
```suggestion
        .then(({height, width}) => 
```

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -230,158 +225,170 @@ class AvatarWithImagePicker extends React.Component {
+                    originalFileName={originalFileName}
+                    fallbackSource={fallbackIcon}
+                >
+                    {({ show }) => (
```
```suggestion
                    {({show}) => (
```

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -230,158 +225,170 @@ class AvatarWithImagePicker extends React.Component {
+                >
+                    {({ show }) => (
+                        <AttachmentPicker>
+                            {({ openPicker }) => {
```
```suggestion
                            {({openPicker}) => {
```

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -230,158 +225,170 @@ class AvatarWithImagePicker extends React.Component {
+            {errorData.validationError && (
+                <DotIndicatorMessage
+                    style={[styles.mt6]}
+                    messages={{ 0: translate(errorData.validationError, errorData.phraseParam) }}
```
```suggestion
                    messages={{0: translate(errorData.validationError, errorData.phraseParam)}}
```

eVoloshchak commented:
> The file has been already reverted to the original :)

it's still present in the list of the changed files, there should be only 2 files changed
<img width="333" alt="image" src="https://github.com/Expensify/App/assets/9059945/7e2ee4b3-13e5-4360-a3cd-8741882290ac">

keisyrzk committed:
prettier

keisyrzk committed:
restore podfile.lock from main

keisyrzk commented:
> > The file has been already reverted to the original :)
> 
> it's still present in the list of the changed files, there should be only 2 files changed <img alt="image" width="333" src="https://user-images.githubusercontent.com/9059945/276738429-7e2ee4b3-13e5-4360-a3cd-8741882290ac.png">

I have reverted it once again. Looks like Podfile.lock disappears from the change map.

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -54,9 +52,6 @@ const propTypes = {
     }).isRequired,
 
-    /** Flag to see if image is being uploaded */
-    isUploading: PropTypes.bool,
```
If we remove `isUploading` prop, we also shouldn't pass it to `AvatarWithImagePicker` from `WorkspaceSettingsPage`

keisyrzk committed:
merge main + conflicts

keisyrzk committed:
pr comments

keisyrzk commented:
@eVoloshchak updates ready and pushed

keisyrzk committed:
updates - linter and PR comments

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -94,7 +89,7 @@ const propTypes = {
     isFocused: PropTypes.bool.isRequired,
 
-    ...withLocalizePropTypes,
+    ...withNavigationFocusPropTypes,
```
`withNavigationFocusPropTypes` doesn't exist
We already have `isFocused` on line 90

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -232,158 +224,168 @@ class AvatarWithImagePicker extends React.Component {
+    };
+
+    return (
+        <View style={StyleSheet.flatten([styles.alignItemsCenter, style])}>
```
Please re-test this on all of the platforms after making the final changes, currently this is crashing (`StyleSheet` isn't imported)

https://github.com/Expensify/App/assets/9059945/fe709ecc-fc8a-4f50-9439-3ccf21706e08

keisyrzk replied:
@eVoloshchak please find the videos below

macOS desktop
https://github.com/Expensify/App/assets/18525360/90b71053-79ca-4ad8-a22b-d3441bcbf5fe

macOS safari
https://github.com/Expensify/App/assets/18525360/76af6956-a06f-4358-bc04-e1af0b22dfd0

macOS chrome
https://github.com/Expensify/App/assets/18525360/c705c44e-3769-4663-847c-5f7719ec4890

ios
https://github.com/Expensify/App/assets/18525360/80af5907-d25f-430d-b0e1-a6cb0bb68bb4

ios safari
https://github.com/Expensify/App/assets/18525360/c78d8de8-6bb0-436e-96b6-ec803a32bd8f

android
https://github.com/Expensify/App/assets/18525360/9284b056-92fb-4299-9496-713efa8ef43b

android chrome
https://github.com/Expensify/App/assets/18525360/38c8efc3-9683-45cb-acae-18e8698f7d56

keisyrzk committed:
updates PR

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -91,10 +86,7 @@ const propTypes = {
     originalFileName: PropTypes.string,
 
-    /** Whether navigation is focused */
-    isFocused: PropTypes.bool.isRequired,
```
It should be the other way around
We need to keep the `isFocused` prop yype, but remove the `withNavigationFocusPropTypes` (it doesn't exist)

keisyrzk committed:
keep isFocused

keisyrzk commented:
@eVoloshchak ok got it, I have updated the code for `isFocused` and pushed the changes.

eVoloshchak commented:
@keisyrzk
There's a linter error now
<img width="573" alt="Screenshot 2023-10-27 at 18 21 23" src="https://github.com/Expensify/App/assets/9059945/1555661d-3017-40fc-893a-32500dcb9278">

keisyrzk committed:
merge conflicts

keisyrzk committed:
merge lint issues

keisyrzk commented:
@eVoloshchak I have merged the newest main and resolved conflicts. I did also fix the linter error and warnings.

eVoloshchak commented:
@keisyrzk, the order of imports should be different, the Lint check will still fail 
run `npm run prettier` and commit the changes

keisyrzk committed:
linter prettier fix

keisyrzk commented:
@eVoloshchak ran prettier and linter again and pushed updates

eVoloshchak reviewed (changes requested):
`SpinningIndicatorAnimation` isn't deleted anymore, must have been a mismatch during merge

keisyrzk committed:
retrieve animation

keisyrzk commented:
@eVoloshchak ok I have added the `SpinningIndicatorAnimation`

blazejkustra commented:
@keisyrzk @eVoloshchak Why `SpinningIndicator` was restored? It wasn't used before, [context](https://github.com/Expensify/App/pull/28625#discussion_r1345864574)

eVoloshchak commented:
>  ok I have added the SpinningIndicatorAnimation

@keisyrzk, I meant we need to delete the `SpinningIndicatorAnimation.js` file

deetergp commented:
…and resolve the merge conflict 😬

keisyrzk committed:
delete SpinningIndicatorAnimation

keisyrzk commented:
@eVoloshchak oh ok, I misunderstood your comment. I have removed the `SpinningIndicatorAnimation.js` file and the animation code I added last time. Changes pushed.

keisyrzk committed:
Merge branch 'main' into update/AvatarWithImagePicker_16119

eVoloshchak reviewed (changes requested):
**Bug:** 
when selecting an image, you can select other files (pdf, mp4, etc. files are not greyed out)
<img width="248" alt="Screenshot 2023-11-08 at 16 49 42" src="https://github.com/Expensify/App/assets/9059945/a317642d-004d-4731-abf0-ca5efb76e6fe">

Here's how it looks on the current main, only images can be selected
<img width="234" alt="Screenshot 2023-11-08 at 16 48 26" src="https://github.com/Expensify/App/assets/9059945/4b28d530-8e80-4941-92c9-5b330ee469d3">

eVoloshchak commented on src/components/AvatarWithImagePicker.js:
```diff
@@ -232,158 +232,168 @@ class AvatarWithImagePicker extends React.Component {
+                    fallbackSource={fallbackIcon}
+                >
+                    {({show}) => (
+                        <AttachmentPicker>
```
Is there a reason this was omitted?

keisyrzk replied:
I believe it is because the weird conflicts with the main I had to resolve

keisyrzk committed:
Update src/components/AvatarWithImagePicker.js

Co-authored-by: Eugene Voloshchak <copyreading@gmail.com>

eVoloshchak reviewed (changes requested):
Crash on Android

https://github.com/Expensify/App/assets/9059945/955fdedb-890c-4e3c-b520-76ece953d51a

koko57 committed:
Merge branch 'main' into update/AvatarWithImagePicker_16119

koko57 committed:
fix: restore role instead of accesbilityRole to prevent crash on android

keisyrzk committed:
Merge branch 'main' into update/AvatarWithImagePicker_16119

eVoloshchak commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

eVoloshchak reviewed (approved):
LGTM

deetergp reviewed (commented):
LGTM2! Nice to see this one get over the finish line. Thanks for the hard work @keisyrzk & @eVoloshchak!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6951287204) to staging by https://github.com/deetergp in version: 1.4.2-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6972415425) to production by https://github.com/luacmartins in version: 1.4.2-3 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
